// vite-plugin-json-server/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default,
  pluginJsonServer: () => pluginJsonServer
});
module.exports = __toCommonJS(src_exports);
var import_json_server = __toESM(require("json-server"), 1);
var import_path = require("path");
var import_is = __toESM(require("json-server/lib/cli/utils/is"), 1);
var import_load = __toESM(require("json-server/lib/cli/utils/load"), 1);
var import_connect_pause = __toESM(require("connect-pause"), 1);
var assertPluginOptions = (pluginOptions) => {
  const {
    apiPath = "/api",
    profile = "",
    source = "db.json",
    static: staticDir = "public",
    // routes = "routes.json",
    // snapshots = "snapshots",
    // middlewares = [],
    unwatch = false,
    readOnly = false,
    bodyParser = true,
    noCors = false,
    noGzip = false,
    delay = 10,
    id = "_id",
    foreignKeySuffix = "_id",
    quiet = false
  } = pluginOptions;
  return {
    apiPath,
    profile,
    source: import_is.default.URL(source) ? source : (0, import_path.join)(profile, source),
    static: (0, import_path.join)(profile, staticDir),
    // routes: join(profile, staticDir),
    // snapshots: join(profile, snapshots),
    // middlewares: join(profile, middlewares),
    unwatch,
    readOnly,
    bodyParser,
    noCors,
    noGzip,
    delay,
    id,
    foreignKeySuffix,
    quiet
  };
};
var pluginJsonServer = (userOptions = {}) => {
  const opts = assertPluginOptions(userOptions);
  const createServer = async (config) => {
    const { root, logger } = config;
    try {
      const routeOpts = {
        id: opts.id,
        foreignKeySuffix: opts.foreignKeySuffix
      };
      const defaultsOpts = {
        static: (0, import_path.join)(root, opts.static),
        logger: !opts.quiet,
        readOnly: opts.readOnly,
        noCors: opts.noCors,
        noGzip: opts.noGzip,
        bodyParser: opts.bodyParser
      };
      logger.info("  Loading: " + opts.source);
      const db = await (0, import_load.default)(opts.source);
      const app = import_json_server.default.create();
      const router = import_json_server.default.router(db, routeOpts);
      const defaults = import_json_server.default.defaults(defaultsOpts);
      app.use(defaults);
      if (opts.delay) {
        logger.info("  Delay: " + opts.delay);
        app.use((0, import_connect_pause.default)(opts.delay));
      }
      app.use(router);
      return app;
    } catch (error) {
      console.log("createServer Error:", error);
      return void 0;
    }
  };
  return {
    name: "vite-plugin-json-server",
    configureServer: async (devServer) => {
      const { config, watcher } = devServer;
      let app = await createServer(config);
      if (!opts.unwatch) {
        const files = [opts.source];
        watcher.on("change", async (file) => {
          const reload = files.find((it) => file.endsWith(it));
          if (reload) {
            console.error(`Change file '${file}'`);
            app = await createServer(config);
          }
        });
      }
      devServer.middlewares.use(
        opts.apiPath,
        (req, res, next) => {
          app && app(req, res, next);
        }
      );
    }
  };
};
var src_default = pluginJsonServer;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  pluginJsonServer
});
