// vite-plugin-json-server/dist/index.js
// src/index.ts
import jsonServer from "json-server";
import { join } from "path";
import is from "json-server/lib/cli/utils/is";
import load from "json-server/lib/cli/utils/load";
import pause from "connect-pause";
var assertPluginOptions = (pluginOptions) => {
  const {
    apiPath = "/api",
    profile = "",
    source = "db.json",
    static: staticDir = "public",
    // routes = "routes.json",
    // snapshots = "snapshots",
    // middlewares = [],
    unwatch = false,
    readOnly = false,
    bodyParser = true,
    noCors = false,
    noGzip = false,
    delay = 10,
    id = "_id",
    foreignKeySuffix = "_id",
    quiet = false
  } = pluginOptions;
  return {
    apiPath,
    profile,
    source: is.URL(source) ? source : join(profile, source),
    static: join(profile, staticDir),
    // routes: join(profile, staticDir),
    // snapshots: join(profile, snapshots),
    // middlewares: join(profile, middlewares),
    unwatch,
    readOnly,
    bodyParser,
    noCors,
    noGzip,
    delay,
    id,
    foreignKeySuffix,
    quiet
  };
};
var pluginJsonServer = (userOptions = {}) => {
  const opts = assertPluginOptions(userOptions);
  const createServer = async (config) => {
    const { root, logger } = config;
    try {
      const routeOpts = {
        id: opts.id,
        foreignKeySuffix: opts.foreignKeySuffix
      };
      const defaultsOpts = {
        static: join(root, opts.static),
        logger: !opts.quiet,
        readOnly: opts.readOnly,
        noCors: opts.noCors,
        noGzip: opts.noGzip,
        bodyParser: opts.bodyParser
      };
      logger.info("  Loading: " + opts.source);
      const db = await load(opts.source);
      const app = jsonServer.create();
      const router = jsonServer.router(db, routeOpts);
      const defaults = jsonServer.defaults(defaultsOpts);
      app.use(defaults);
      if (opts.delay) {
        logger.info("  Delay: " + opts.delay);
        app.use(pause(opts.delay));
      }
      app.use(router);
      return app;
    } catch (error) {
      console.log("createServer Error:", error);
      return void 0;
    }
  };
  return {
    name: "vite-plugin-json-server",
    configureServer: async (devServer) => {
      const { config, watcher } = devServer;
      let app = await createServer(config);
      if (!opts.unwatch) {
        const files = [opts.source];
        watcher.on("change", async (file) => {
          const reload = files.find((it) => file.endsWith(it));
          if (reload) {
            console.error(`Change file '${file}'`);
            app = await createServer(config);
          }
        });
      }
      devServer.middlewares.use(
        opts.apiPath,
        (req, res, next) => {
          app && app(req, res, next);
        }
      );
    }
  };
};
var src_default = pluginJsonServer;
export {
  src_default as default,
  pluginJsonServer
};
